Ignore identical types when loading routable components

Silverdark opened the pull request:
# Ignore identical types when loading routable components

<!-- Thank you for submitting a pull request to our repo. -->

<!-- If this is your first PR in the ASP.NET Core repo, please run through the checklist
below to ensure a smooth review and merge process for your PR. -->

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

<!-- Once all that is done, you're ready to go. Open the PR with the content below. -->

Summary of the changes (Less than 80 chars)

## Description

Instead of using a `List` of types, I replaced it with a `HashSet` so that no duplicate entries are allowed when the routable components are loaded.

This issue does happen, if you specify the same assembly for the `AppAssembly` and the `AdditionalAssemblies`. It does not happen if you specify the same assembly multiple times for `AdditionalAssemblies`, as this is already a `HashSet`.

Fixes #38882

Silverdark committed:
Ignore identical types when loading routable components

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=39129) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=39129) before we can accept your contribution.<br/><br/>:x: Silverdark [sign now](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=39129)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/aspnetcore?pullRequest=39129) it.</sub>

Silverdark commented:
Just checked the failing test actions

> Using member 'System.Reflection.Assembly.ExportedTypes.get' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. Types might be removed.

I really don't know why, but this occurs only when using the `HashSet`. When I replace it with a simple `Contains` condition, the tests are working fine (locally). For me this is strange, as in both cases `assembly.ExportedTypes` will be used.

```csharp
foreach (var type in assembly.ExportedTypes)
{
    if (typeof(IComponent).IsAssignableFrom(type) && type.IsDefined(typeof(RouteAttribute)))
    {
        if (!routeableComponents.Contains(type)) // <-- This part here
        {
            routeableComponents.Add(type);
        }
    }
}
```

Silverdark committed:
Fixed test and revert HashSet to List in order to mitigate "RequiresUnreferencedCodeAttribute" error

pranavkm commented on src/Components/Components/src/Routing/RouteTableFactory.cs:
```diff
@@ -56,7 +56,10 @@ static void GetRouteableComponents(List<Type> routeableComponents, Assembly asse
                 if (typeof(IComponent).IsAssignableFrom(type) && type.IsDefined(typeof(RouteAttribute)))
                 {
-                    routeableComponents.Add(type);
+                    if (!routeableComponents.Contains(type))
```
This is going to do a linear scan isn't it? Could you tell me what the error was when you used a HashSet instead? 

An alternative fix is to de-dupe AppAssembly and AdditionalAssemblies. Essentially you could check on line 44 if `assembly != routeKey.AppAssembly` before calling `GetRouteableComponents`

Silverdark replied:
> Could you tell me what the error was when you used a HashSet instead?

The tests are failing with the following reason ([example in Azure DevOps](https://dev.azure.com/dnceng/public/_build/results?buildId=1521937&view=logs&j=5ac7b393-e840-5549-7fb4-a4479af8e7e3&t=29df2fa2-0d20-51bd-e85a-8b546e86c529&l=1506)):

> Using member 'System.Reflection.Assembly.ExportedTypes.get' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. Types might be removed.

I just replaced the `List` with a `HashSet` and adjusted other references in the [commit](https://github.com/dotnet/aspnetcore/pull/39129/commits/bbb9dbe1ef7085e2e820b4d4d51045398749ed5f), but it was failing.

> An alternative fix is to de-dupe AppAssembly and AdditionalAssemblies.

This would also be possible, will change the PR to that solution.

Silverdark committed:
Replace contains with simple condition

Instead of iterating over the list of types all the time, do a simple check if the additional assembly is the same as the app assembly and skip it.

pranavkm commented:
Thanks!
